### PR TITLE
[network] Added caching to node server

### PR DESCRIPTION
**Pull request by JackyWYX**
## Issue

Added caching mechanism to node handling sync traffic for two requests:

1. `DownloaderRequest_BLOCKHEADER`
2. `DownloaderRequest_BLOCK`

The two requested are heavily handled in explorer node. Caching these two requests will save one rlp CPU time each request.

## Test

Code passed local test. 

[test log](https://gist.github.com/JackyWYX/c1e21f37ec4c0cdb8565f03d9b16f8c7)

After changed the solution from `singleflight` to `lru.Cache`, tested again and passed.

[test log](https://gist.github.com/JackyWYX/fa6a57acc78893ee7c8c3e9e9caf2629)

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    **NO**

### CODE REVIEW

File: node/node_syncing.go [hunk 1]
```diff
@@ -477,3 +469,32 @@ func (node *Node) CalculateResponse(request *downloader_pb.DownloaderRequest, in
...
+	errBlockNotExist  = errors.New("block not exist")
+)
+
+func (node *Node) getEncodedBlockHeaderByHash(hash common.Hash) ([]byte, error) {
+	key := fmt.Sprintf("%x", hash)
+	b, err, _ := node.headerGroup.Do(key, func() (interface{}, error) {
+		h := node.Blockchain().GetHeaderByHash(hash)
+		if h == nil {
+			return nil, errHeaderNotExist
+		}
```
Comment (rlan35) [inline node/node_syncing.go:484]: There is no clearing of the cache?

Comment (JackyWYX) [reply node/node_syncing.go:484]: Right! Thanks for pointing that out!

File: node/node_syncing.go [hunk 2]
```diff
@@ -477,3 +470,49 @@ func (node *Node) CalculateResponse(request *downloader_pb.DownloaderRequest, in
 	}
 	return response, nil
 }
+
+const (
+	headerCacheSize = 50000000
```
Comment (rlan35) [inline node/node_syncing.go:475]: This number is not not size in byte. It's number of items.

Comment (JackyWYX) [reply node/node_syncing.go]: Fixed. Thanks!

Comment (rlan35) [reply node/node_syncing.go]: Should be 10000

Comment (rlan35) [reply node/node_syncing.go]: Should be 10000

**Comment by rlan35** [comment]
After comment fixed and tested. We are good to go.

**Comment by JackyWYX** [comment]
After changed the code, tested again locally and passed.

**Pull request merged**